Mathlib/Topology/Order/HullKernel.lean

```diff
@@ -0,0 +1,250 @@
+
+/- Every relative-closed set of the form `T ↓∩ (↑(upperClosure F))` for `F` finite is a
+relative-closed set of the form `hull T a` where `a = ⨅ F`. -/
+open Finset in
+lemma preimage_upperClosure_finset (hT : ∀ p ∈ T, InfPrime p) (F : Finset α) :
+    T ↓∩ upperClosure F.toSet = hull T (inf F id) := by
+  rw [coe_upperClosure]
+  induction' F using Finset.induction_on with a F' _ I4
+  · simp only [coe_empty, mem_empty_iff_false, iUnion_of_empty, iUnion_empty, Set.preimage_empty,
+      inf_empty, Set.Ici_top]
+    symm
+    by_contra hf
+    rw [← Set.not_nonempty_iff_eq_empty, not_not] at hf
+    obtain ⟨x, hx⟩ := hf
+    exact (hT x (Subtype.coe_prop x)).1 (isMax_iff_eq_top.mpr (eq_top_iff.mpr hx))
+  · simp only [coe_insert, mem_insert_iff, mem_coe, iUnion_iUnion_eq_or_left, Set.preimage_union,
+      preimage_iUnion, inf_insert, id_eq, ← (hull_union_hull hT), ← I4]
+
+/- Every relative-open set of the form `T ↓∩ (↑(upperClosure F))ᶜ` for `F` finite is a relative-open
+set of the form `(hull T a)ᶜ` where `a = ⨅ F`. -/
+open Finset in
+lemma preimage_upperClosure_compl_finset (hT : ∀ p ∈ T, InfPrime p) (F : Finset α) :
+    T ↓∩ (upperClosure F.toSet)ᶜ = (hull T (inf F id))ᶜ := by
+  rw [Set.preimage_compl, (preimage_upperClosure_finset hT)]
+
+variable [TopologicalSpace α] [IsLower α]
+
+/-
+The relative-open sets of the form `(hull T a)ᶜ` for `a` in `α` form a basis for the relative
+Lower topology.
+-/
+lemma relativeLowerIsTopologicalBasis (hT : ∀ p ∈ T, InfPrime p) :
+    IsTopologicalBasis { S : Set T | ∃ (a : α), (hull T a)ᶜ = S } := by
+  convert isTopologicalBasis_subtype Topology.IsLower.isTopologicalBasis T
+  ext R
+  simp only [preimage_compl, mem_setOf_eq, IsLower.lowerBasis, mem_image, exists_exists_and_eq_and]
+  constructor <;> intro ha
+  · obtain ⟨a, ha'⟩ :=  ha
+    use {a}
+    rw [← (Function.Injective.preimage_image Subtype.val_injective R), ← ha']
+    simp only [finite_singleton, upperClosure_singleton, UpperSet.coe_Ici, image_val_compl,
+      Subtype.image_preimage_coe, diff_self_inter, preimage_diff, Subtype.coe_preimage_self,
+      true_and]
+    exact compl_eq_univ_diff (Subtype.val ⁻¹' Ici a)
+  · obtain ⟨F, hF⟩ := ha
+    lift F to Finset α using hF.1
+    use Finset.inf F id
+    rw [← (preimage_upperClosure_finset hT), ← hF.2]
+    rfl
+
+end PrimitiveSpectrum
+
+end SemilatticeInf
+
+section PrimativeSpectrum
+
+variable [CompleteLattice α]
+
+variable {T : Set α}
+
+namespace PrimitiveSpectrum
+
+lemma sInter_hull (S : Set α) : ⋂₀ { hull T a | a ∈ S } = hull T (sSup S) := by
+  ext x : 1
+  simp_all only [mem_sInter, mem_setOf_eq, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂,
+    mem_preimage, mem_Ici, sSup_le_iff]
+
+/- When `α` is complete, the relative basis for the lower topology is also closed under arbitrary
+unions. -/
+lemma sUnion_hull_compl (S : Set α) : ⋃₀ { (hull T a)ᶜ | a ∈ S } = (hull T (sSup S))ᶜ := by
+  simp [sUnion_eq_compl_sInter_compl, ← sInter_hull, compl_sInter]
```

Comment:
This lemma looks weird and I would drop it for the same reasons as `hull_compl_inter_hull_compl`